x-pack/platform/plugins/shared/streams/server/lib/streams/attachments/attachment_client.ts

```diff
@@ -0,0 +1,447 @@
+    const filter: QueryDslQueryContainer[] = [{ terms: { [STREAM_NAMES]: [streamName] } }];
+    if (attachmentType) {
+      filter.push({ terms: { [ATTACHMENT_TYPE]: [attachmentType] } });
+    }
+    const attachmentsResponse = await this.clients.storageClient.search({
+      size: 10_000,
+      track_total_hits: false,
+      query: {
+        bool: {
+          filter,
+        },
+      },
+    });
+
+    const attachmentDocuments: AttachmentDocument[] = attachmentsResponse.hits.hits.map(
+      ({ _source }) => _source
+    );
+
+    const attachmentIdsByType: Record<AttachmentType, string[]> = {
+      dashboard: [],
+      rule: [],
+    };
+
+    for (const attachment of attachmentDocuments) {
+      attachmentIdsByType[attachment[ATTACHMENT_TYPE]].push(attachment[ATTACHMENT_ID]);
+    }
+
+    const attachments: Attachment[] = (
+      await Promise.all(
+        ATTACHMENT_TYPES.map(async (type) => {
+          const ids = attachmentIdsByType[type];
+          if (ids.length === 0) {
+            return [];
+          }
+
+          return this.getAttachmentEntitiesMap[type](ids);
+        })
+      )
+    ).flat();
+
+    return attachments;
+  }
+
+  async getSuggestions({
+    query,
+    attachmentTypes,
+    tags,
+  }: {
+    query: string;
+    attachmentTypes?: AttachmentType[];
+    tags?: string[];
+  }): Promise<{ hasMore: boolean; attachments: Attachment[] }> {
+    const perPage = 101;
+
+    const searchAll = !attachmentTypes;
+
+    const searchDashboards = searchAll || attachmentTypes.includes('dashboard');
+    const searchRules = searchAll || attachmentTypes.includes('rule');
+    const suggestionsPromises: Promise<Attachment[]>[] = [];
+    if (searchDashboards) {
+      suggestionsPromises.push(
+        this.getSuggestedEntitiesMap.dashboard({
+          query,
+          tags,
+          perPage,
+        })
+      );
+    }
+
+    if (searchRules) {
+      suggestionsPromises.push(
+        this.getSuggestedEntitiesMap.rule({
+          query,
+          tags,
+          perPage,
+        })
+      );
+    }
+    const suggestions = (await Promise.all(suggestionsPromises)).flat();
+
+    return {
+      attachments: suggestions,
+      hasMore: suggestions.length > perPage - 1,
+    };
+  }
+
+  async syncAttachmentList(
+    streamName: string,
+    links: AttachmentLink[],
+    attachmentType?: AttachmentType
+  ): Promise<{ deleted: AttachmentLink[]; indexed: AttachmentLink[] }> {
+    const filter: QueryDslQueryContainer[] = [{ terms: { [STREAM_NAMES]: [streamName] } }];
+    if (attachmentType) {
+      filter.push({ terms: { [ATTACHMENT_TYPE]: [attachmentType] } });
+    }
+    const attachmentsResponse = await this.clients.storageClient.search({
+      size: 10_000,
+      track_total_hits: false,
+      query: {
+        bool: {
+          filter,
+        },
+      },
+    });
+
+    const existingAssetLinks: AttachmentLink[] = attachmentsResponse.hits.hits.map(
+      ({ _source }) => {
```

Comment:
Do we need to use _source here? I'm always nervous when I see the use of _source as in some places it starts to go away.